android/app/src/main/java/network/perun/app/MainActivity.java

```diff
@@ -0,0 +1,170 @@
+
+            // Create a wallet.
+            Wallet wallet = Prnm.newWallet(ksPath, password);
+            // Import the secret key.
+            Address onChain = wallet.importAccount(sk);
+            Log.i("prnm", "Address: " +onChain.toHex());
+            // 10.0.2.2 is the IP of the host PC when using Android Simulator and the host is running a ganache-cli.
+            // 8545 is the standart port of ganache-cli.
+            String ethUrl = "ws://10.0.2.2:8545";
+
+            // Using null as either Adjudicator or AssetHolder tells the Client to deploy the contracts,
+            // in this case we already deployed them and enter their addresses.
+            Address adjudicator = new Address("0xDc4A7e107aD6dBDA1870df34d70B51796BBd1335");
+            Address assetholder = new Address("0xb051EAD0C6CC2f568166F8fEC4f07511B88678bA");
+            // We will be listening on 127.0.0.1:5750 for new channel proposals with the alias "Alice".
+            Config cfg = new Config("Alice ", onChain, adjudicator, assetholder, dbPath, ethUrl, "127.0.0.1", 5750);
+            node = new Node(cfg, wallet);
+
+            // (Optional) Propose a channel to bob:
+            // (Without the following lines, the node will still accept incoming channel proposals.)
+            //
+            // PerunId (currently an Address) of the peer that we want to open a channel with.
+            Address bob = new Address("0xA298Fc05bccff341f340a11FffA30567a00e651f");
+            // Create the initial balances of the channel, we start with 2000 and bob with 1000.
+            BigInts initBals = Prnm.newBalances(new BigInt(2000), new BigInt(1000));
+            // Send the proposal. Once the channel was accepted, the node will handle the
+            // channel updates and onChain watching.
+            node.propose(bob, initBals, "10.0.2.2", 5750);
+        } catch (Exception e) {
+            Log.e("prnm", e.toString());
+        }
+    }
+}
+
+class Node implements ChannelAcceptor {
+    Client client;
+
+    public Node(Config cfg, Wallet wallet) throws Exception {
+        // Possibly has to deploy contracts, so give it some extra time.
+        Context ctx = Prnm.contextWithTimeout(30);
+        try {
+            client = new Client(ctx, cfg, wallet);
+        } finally {
+            ctx.cancel();
+        }
+
+        // Start a new thread for handling channel proposals.
+        new Thread(() -> {
+            ProposalHandler propHandler = new ProposalHandler(this);
+            client.handleChannelProposals(propHandler);
+        }).start();
+    }
+
+    public void propose(Address peer, BigInts initBals, String ip, int port) throws Exception {
+        // This is safe to call more than once.
+        client.addPeer(peer, ip, port);
+        // Has to send transactions, so give it some extra time.
+        Context ctx = Prnm.contextWithTimeout(30);
+        try {
+            accept(client.proposeChannel(ctx, peer, 300, initBals));
```

Comment:
When proposing a channel, the context needs to allow for twice the challenge duration to pass, plus extra time for up to three transactions. This is because if a peer fails funding, there still needs to be enough time in the context to refund the channel with a force-close. So better lower the challengeduration to, say, 60 sec, and increase proposal timeout to 600 sec.